### PR TITLE
Make game controller state consistent

### PR DESCRIPTION
**Pull request by ichbinJosi**
## Introduced Changes

Implements a consistently used `FilteredGameStateController` to enable a single layer abstraction architecture.

## How to Test

Check if everything still works as expected regarding anything concerned with `GameStates` 🤡

### CODE REVIEW

File: crates/types/src/filtered_game_controller_state.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#[derive(Default, Clone, Copy, Debug, Serialize, Deserialize, SerializeHierarchy)]
+pub struct FilteredGameControllerState {
+    pub game_state: FilteredGameState,
+    pub opponent_game_state: FilteredGameState,
+    pub game_phase: GamePhase,
+    pub kicking_team: Team,
+    pub penalties: Players<Option<Penalty>>,
+    pub remaining_amount_of_messages: u16,
+    pub sub_state: Option<SubState>,
+    pub hulks_team_is_home_after_coin_toss: bool,
```
Comment (schmidma) [inline crates/types/src/filtered_game_controller_state.rs:16]:
```suggestion
    pub own_team_is_home_after_coin_toss: bool,
```

File: crates/types/src/lib.rs [hunk 2]
```diff
@@ -17,7 +17,8 @@ pub mod field_border;
 pub mod field_color;
 pub mod field_dimensions;
 pub mod field_marks;
-pub mod filtered_game_state;
+pub mod filtered_game_controller_state;
+pub mod filtered_game_states;
```
Comment (schmidma) [inline crates/types/src/lib.rs:21]:
```suggestion
pub mod filtered_game_state;
```

File: etc/parameters/default.json [hunk 3]
```diff
@@ -1167,5 +1167,6 @@
       "knee_pitch": 0.0
     }
   },
-  "angular_velocity_smoothing_factor": 0.1
+  "angular_velocity_smoothing_factor": 0.1,
+  "center_circle_obstacle_increasement": 1.2
```
Comment (schmidma) [inline etc/parameters/default.json:1171]:
```suggestion
  "center_circle_obstacle_increase": 1.2
```

File: etc/parameters/head.P0000074A06S9A900031.json [hunk 4]
```diff
@@ -15,4 +15,4 @@
       ]
     }
   }
-}
+}
```
Comment (schmidma) [inline etc/parameters/head.P0000074A06S9A900031.json:18]: missing newline

File: etc/parameters/nao_location [hunk 5]
```diff
@@ -1 +1 @@
-smd
\ No newline at end of file
+leipzig
```
Comment (schmidma) [inline etc/parameters/nao_location:1]: leave that as `smd`

File: crates/control/src/behavior/intercept_ball.rs [hunk 6]
```diff
@@ -18,25 +18,31 @@ pub fn execute(
...
-    ) = world_state.game_controller_state
+    ) = world_state.filtered_game_controller_state
     {
         return None;
     }
     match (
-        world_state.filtered_game_state,
+        world_state
+            .filtered_game_controller_state
+            .map(|filtered_game_controller_state| filtered_game_controller_state.game_state),
```
Comment (schmidma) [inline crates/control/src/behavior/intercept_ball.rs:36]: please extract that into a variable before the `match`

File: crates/control/src/behavior/node.rs [hunk 7]
```diff
@@ -145,22 +145,29 @@ impl Behavior {
             Role::MidfielderRight => actions.push(Action::SupportRight),
             Role::ReplacementKeeper => actions.push(Action::DefendGoal),
             Role::Searcher => actions.push(Action::Search),
-            Role::Striker => match world_state.filtered_game_state {
-                None | Some(FilteredGameState::Playing { ball_is_free: true }) => {
+            Role::Striker => match world_state
+                .filtered_game_controller_state
+                .map(|filtered_game_controller_state| filtered_game_controller_state.game_state)
```
Comment (schmidma) [inline crates/control/src/behavior/node.rs:150]: same here, extract into variable

File: crates/control/src/behavior/node.rs [hunk 8]
```diff
@@ -145,22 +145,29 @@ impl Behavior {
...
+            {
+                None
+                | Some(FilteredGameState::Playing {
+                    ball_is_free: true,
+                    kick_off: _,
```
Comment (schmidma) [inline crates/control/src/behavior/node.rs:155]:
```suggestion
                    ..
```

File: crates/control/src/behavior/support.rs [hunk 9]
```diff
@@ -56,10 +56,14 @@ fn support_pose(
...
     let supporting_position = ball.ball_in_field + offset_vector;
-    let clamped_x = match world_state.filtered_game_state {
+    let clamped_x = match world_state
+        .filtered_game_controller_state
+        .map(|filtered_game_controller_state| filtered_game_controller_state.game_state)
+    {
         Some(FilteredGameState::Ready { .. })
         | Some(FilteredGameState::Playing {
             ball_is_free: false,
+            kick_off: _,
```
Comment (schmidma) [inline crates/control/src/behavior/support.rs:66]:
```suggestion
            ..
```

File: crates/control/src/behavior/support.rs [hunk 10]
```diff
@@ -56,10 +56,14 @@ fn support_pose(
         Side::Right => FRAC_PI_4,
     }) * -(Vector2::x() * distance_to_ball);
     let supporting_position = ball.ball_in_field + offset_vector;
-    let clamped_x = match world_state.filtered_game_state {
+    let clamped_x = match world_state
+        .filtered_game_controller_state
+        .map(|filtered_game_controller_state| filtered_game_controller_state.game_state)
```
Comment (schmidma) [inline crates/control/src/behavior/support.rs:61]: as this is used so often, you may introduce a function for that?

File: crates/control/src/game_controller_state_filter.rs [hunk 11]
```diff
@@ -50,59 +48,101 @@ impl GameStateFilter {
...
 
-        let ball_detected_far_from_kick_off_point = context
-            .ball_position
-            .map(|ball| {
-                let absolute_ball_position = *context.robot_to_field * ball.position;
-                distance(&absolute_ball_position, &Point2::origin())
-                    > context.config.distance_to_consider_ball_moved_in_kick_off
-            })
-            .unwrap_or(false);
+pub struct FilteredGameStates {
```
Comment (schmidma) [inline crates/control/src/game_controller_state_filter.rs:82]: does this need to be public?

File: crates/control/src/game_controller_state_filter.rs [hunk 12]
```diff
@@ -50,59 +48,101 @@ impl GameStateFilter {
...
 
-        let filtered_game_state = self.state.construct_hulks_filtered_game_state(
-            context.game_controller_state,
-            context.cycle_time.start_time,
-            ball_detected_far_from_kick_off_point,
-            context.config,
-        );
+#[allow(clippy::too_many_arguments)]
+pub fn filter_game_states(
+    robot_to_field: &mut Isometry2<f32>,
```
Comment (schmidma) [inline crates/control/src/game_controller_state_filter.rs:89]: This should better not be mutable?!

**Review by schmidma — CHANGES_REQUESTED** [review]
only some minor stuff

File: crates/types/src/filtered_game_controller_state.rs [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+use crate::{filtered_game_states::FilteredGameState, players::Players};
+
+#[derive(Default, Clone, Copy, Debug, Serialize, Deserialize, SerializeHierarchy)]
+pub struct FilteredGameControllerState {
+    pub game_state: FilteredGameState,
+    pub opponent_game_state: FilteredGameState,
+    pub game_phase: GamePhase,
+    pub kicking_team: Team,
+    pub penalties: Players<Option<Penalty>>,
+    pub remaining_amount_of_messages: u16,
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/types/src/filtered_game_controller_state.rs:14]: Done

File: crates/types/src/filtered_game_controller_state.rs (hunk 1 above)
Comment (schluis) [reply crates/types/src/filtered_game_controller_state.rs:16]: Done

File: crates/types/src/lib.rs (hunk 2 above)
Comment (schluis) [reply crates/types/src/lib.rs:21]: Done

File: etc/parameters/default.json (hunk 3 above)
Comment (schluis) [reply etc/parameters/default.json:1171]: Done

File: etc/parameters/head.P0000074A06S9A900031.json (hunk 4 above)
Comment (schluis) [reply etc/parameters/head.P0000074A06S9A900031.json:18]: Done

File: etc/parameters/nao_location (hunk 5 above)
Comment (schluis) [reply etc/parameters/nao_location:1]: Done

File: crates/control/src/behavior/intercept_ball.rs (hunk 6 above)
Comment (schluis) [reply crates/control/src/behavior/intercept_ball.rs:36]: As discussed postponed.

File: crates/control/src/behavior/intercept_ball.rs [hunk 14]
```diff
@@ -18,25 +18,31 @@ pub fn execute(
...
         world_state.ball,
         world_state.robot.robot_to_field,
     ) {
         (
-            Some(FilteredGameState::Playing { ball_is_free: true }) | None,
+            Some(FilteredGameState::Playing {
+                ball_is_free: true,
+                kick_off: _,
```
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/behavior/intercept_ball.rs:43]: Done

File: crates/control/src/behavior/node.rs (hunk 7 above)
Comment (schluis) [reply crates/control/src/behavior/node.rs:150]: Also postponed

File: crates/control/src/behavior/node.rs (hunk 8 above)
Comment (schluis) [reply crates/control/src/behavior/node.rs:155]: Done

File: crates/control/src/behavior/support.rs (hunk 9 above)
Comment (schluis) [reply crates/control/src/behavior/support.rs:66]: Done

File: crates/control/src/behavior/support.rs (hunk 10 above)
Comment (schluis) [reply crates/control/src/behavior/support.rs:61]: We will fix the bug in the framework. See Issue #668

File: crates/control/src/game_controller_state_filter.rs (hunk 11 above)
Comment (schluis) [reply crates/control/src/game_controller_state_filter.rs:82]: No, Done

File: crates/control/src/game_controller_state_filter.rs (hunk 12 above)
Comment (schluis) [reply crates/control/src/game_controller_state_filter.rs:89]: Yep. Done

File: crates/control/src/behavior/intercept_ball.rs (hunk 6 above)
Comment (ichbinJosi) [reply crates/control/src/behavior/intercept_ball.rs:36]: Done

File: crates/control/src/behavior/node.rs (hunk 7 above)
Comment (ichbinJosi) [reply crates/control/src/behavior/node.rs:150]: Done

File: crates/control/src/behavior/support.rs (hunk 10 above)
Comment (ichbinJosi) [reply crates/control/src/behavior/support.rs]: Extracted into variable as discussed

File: crates/control/src/game_controller_state_filter.rs [hunk 15]
```diff
@@ -50,59 +48,101 @@ impl GameStateFilter {
...
-            .ball_position
-            .map(|ball| {
-                let absolute_ball_position = *context.robot_to_field * ball.position;
-                distance(&absolute_ball_position, &Point2::origin())
-                    > context.config.distance_to_consider_ball_moved_in_kick_off
-            })
-            .unwrap_or(false);
+struct FilteredGameStates {
+    pub filtered_game_state: FilteredGameState,
+    pub filtered_opponent_game_state: FilteredGameState,
```
Comment (schmidma) [inline crates/control/src/game_controller_state_filter.rs:84]:
```suggestion
    own: FilteredGameState,
    opponent: FilteredGameState,
```

The `pub` might not be needed. Maybe this is a leftover after removing the `pub` for the entire struct? 
Additionally, the field names, can be shortened. Following the "do not repeat yourself" rule, the type should not be repeated in the variable name.

File: crates/control/src/game_controller_state_filter.rs [hunk 16]
```diff
@@ -240,7 +280,7 @@ fn in_kick_off_grace_period(
 }
 
 #[derive(Clone, Copy, Deserialize, Serialize)]
-enum State {
+pub enum State {
```
Comment (schmidma) [inline crates/control/src/game_controller_state_filter.rs:283]:
```suggestion
enum State {
```

File: crates/control/src/game_controller_state_filter.rs (hunk 15 above)
Comment (ichbinJosi) [reply crates/control/src/game_controller_state_filter.rs:84]: Done

File: crates/control/src/game_controller_state_filter.rs (hunk 16 above)
Comment (ichbinJosi) [reply crates/control/src/game_controller_state_filter.rs:283]: Done

**Review by schmidma — APPROVED** [review]

**Pull request merged**